code-dot-org/code-dot-org
# migrate all remaining units to use new seed and edit experience

## Background

Finishes [PLAT-1428]. This step will unblock Dani from being able to make progress on certain self-paced PL work items. 

See [work plan](https://docs.google.com/document/d/1aH_YxmhSXsYcDHmDIU4MDtIu2KoqCgxih6nhGd1SBdo/edit#) for past work which has led up to this step. To summarize:
* migrated all HOC units
* migrated 271 other units
* updated test coverage to operate on migrated units
* fixed several bugs caught by manual and automated testing 

## Description

1. Migrate 140 remaining unmigrated units, including all CSF / CSD / CSP courses from before 2021. Here is what I did:

```
[development] dashboard > puts Script.all.reject(&:is_migrated).map(&:name).sort.map(&:dump).join(',')
"20-hour","2021drafting","AlgebraA","AlgebraB","ai-unit-pilot","algebra","andrea-test","code-break","code-break-younger","contagion-guided-practice","contagion-pilot","course1","course2","course3","course4","coursea-2017","coursea-2018","coursea-2019","coursea-2020","courseb-2017","courseb-2018","courseb-2019","courseb-2020","coursec-2017","coursec-2018","coursec-2019","coursec-2020","coursed-2017","coursed-2018","coursed-2019","coursed-2020","coursee-2017","coursee-2018","coursee-2019","coursee-2020","coursef-2017","coursef-2018","coursef-2019","coursef-2020","csd-post-survey","csd-post-survey-2018","csd1-2017","csd1-2018","csd1-2019","csd1-2020","csd2-2017","csd2-2018","csd2-2019","csd2-2020","csd2-projects-temp","csd3-2017","csd3-2018","csd3-2019","csd3-2020","csd3-virtual","csd4-2017","csd4-2018","csd4-2019","csd4-2020","csd4-preview-2021","csd5-2017","csd5-2018","csd5-2019","csd5-2020","csd5-preview-2021","csd6-2017","csd6-2018","csd6-2019","csd6-2020","csf-2021-pilot","csp-create-2017","csp-create-2018","csp-create-2019","csp-explore-2017","csp-explore-2018","csp-explore-2019","csp-mid-survey","csp-post-survey","csp-post-survey-2018","csp-post-survey-2020","csp-pre-survey","csp1-2017","csp1-2018","csp1-2019","csp1-2020","csp10-2020","csp2-2017","csp2-2018","csp2-2019","csp2-2020","csp3-2017","csp3-2018","csp3-2019","csp3-2020","csp3-research-mxghyt","csp3-staging","csp3-virtual","csp4-2017","csp4-2018","csp4-2019","csp4-2020","csp5-2017","csp5-2018","csp5-2019","csp5-2020","csp5-virtual","csp6-2020","csp7-2020","csp8-2020","csp9-2020","csppostap-2017","csppostap-2018","csppostap-2019","deepdive-debugging","denny-science","denny-science-8","express-2017","express-2018","express-2019","express-2020","haikubot-pilot","jess-test-script","new-d","new-express","new-f","peru-2019","pixelation","pre-express-2017","pre-express-2018","pre-express-2019","pre-express-2020","pwc","time4cs-control-unit-1","time4cs-control-unit-2","time4cs-experiment-unit-1","time4cs-experiment-unit-2","vpl-csd-2020","vpl-csd-2021","vpl-csp-2020","vpl-csp-2021"

bundle exec bin/curriculum/migrate_unit.rb -v -u "20-hour","2021drafting","AlgebraA","AlgebraB","ai-unit-pilot","algebra","andrea-test","code-break","code-break-younger","contagion-guided-practice","contagion-pilot","course1","course2","course3","course4","coursea-2017","coursea-2018","coursea-2019","coursea-2020","courseb-2017","courseb-2018","courseb-2019","courseb-2020","coursec-2017","coursec-2018","coursec-2019","coursec-2020","coursed-2017","coursed-2018","coursed-2019","coursed-2020","coursee-2017","coursee-2018","coursee-2019","coursee-2020","coursef-2017","coursef-2018","coursef-2019","coursef-2020","csd-post-survey","csd-post-survey-2018","csd1-2017","csd1-2018","csd1-2019","csd1-2020","csd2-2017","csd2-2018","csd2-2019","csd2-2020","csd2-projects-temp","csd3-2017","csd3-2018","csd3-2019","csd3-2020","csd3-virtual","csd4-2017","csd4-2018","csd4-2019","csd4-2020","csd4-preview-2021","csd5-2017","csd5-2018","csd5-2019","csd5-2020","csd5-preview-2021","csd6-2017","csd6-2018","csd6-2019","csd6-2020","csf-2021-pilot","csp-create-2017","csp-create-2018","csp-create-2019","csp-explore-2017","csp-explore-2018","csp-explore-2019","csp-mid-survey","csp-post-survey","csp-post-survey-2018","csp-post-survey-2020","csp-pre-survey","csp1-2017","csp1-2018","csp1-2019","csp1-2020","csp10-2020","csp2-2017","csp2-2018","csp2-2019","csp2-2020","csp3-2017","csp3-2018","csp3-2019","csp3-2020","csp3-research-mxghyt","csp3-staging","csp3-virtual","csp4-2017","csp4-2018","csp4-2019","csp4-2020","csp5-2017","csp5-2018","csp5-2019","csp5-2020","csp5-virtual","csp6-2020","csp7-2020","csp8-2020","csp9-2020","csppostap-2017","csppostap-2018","csppostap-2019","deepdive-debugging","denny-science","denny-science-8","express-2017","express-2018","express-2019","express-2020","haikubot-pilot","jess-test-script","new-d","new-express","new-f","peru-2019","pixelation","pre-express-2017","pre-express-2018","pre-express-2019","pre-express-2020","pwc","time4cs-control-unit-1","time4cs-control-unit-2","time4cs-experiment-unit-1","time4cs-experiment-unit-2","vpl-csd-2020","vpl-csd-2021","vpl-csp-2020","vpl-csp-2021"

[development] dashboard > Script.all.reject(&:is_migrated).count
=> 0
```

2. to fix lesson-standards mappings, I used localhost-studio.code.org:3000/admin/standards to import standards for each CSF 2019 and 2020 unit, and ran the following to serialize the changes:
```
Script.all.select {|s| s.curriculum_umbrella == 'CSF' && ['2019','2020'].include?(s.version_year)}.each(&:write_script_json)
```
the results are in their own commit in this PR. see the adhoc doc mentioned below for more details.

## Testing story

Live site test coverage
* Previously shipped migration of ~30 HOC units ~2 months ago. no reported issues.
* Previously shipped migration of 271 other units 2-3 weeks ago, including high-profile /s/hourofcode . no reported issues.

Automated testing (see [work plan](https://docs.google.com/document/d/1aH_YxmhSXsYcDHmDIU4MDtIu2KoqCgxih6nhGd1SBdo/edit#) for details)
* updated many unit tests covering legacy unit seeding functionality to cover migrated seeding functionality instead
* migrated units covered by ui tests, including allthethings, allthehiddenthings, alltheplcthings, allthsurveys, allthettsthings
* migrated units covered by [course_versions.feature](https://github.com/code-dot-org/code-dot-org/blob/d22aed2dfecebff7b4278fd79dde6eb2256eaa3c/dashboard/test/ui/features/learning_platform/course_versions.feature) ui tests

Manual testing on adhoc
* see [adhoc migrate units](https://docs.google.com/document/d/1YGayvgTugZ6UvvC-sEjvWRA9q0YZP3Ymk8c4z2qlqnI/edit#) for details of manual testing

## Deployment strategy

My proposal is to merge this into staging provided that it is ready before the deadline. This isn't 100% necessary, but the main advantage would be that I can then use the lockdown time to start removing the underlying script seed code, which I won't want to do until this change has been in production for a day or two.

The reason this seems safe enough to me is that this feels pretty well tested now, and I also want to trust the process that we've allowed enough buffer time between the Friday lockdown deadline and hour of code launch. if something goes wrong, we can use the remaining days to follow up with fixes.

If this doesn't feel good to reviewers, there's definitely the option of merging this to staging-next instead.

## Follow-up work

Once this reaches production, double-check a few high-profile features, such as:
* standards view is still working
* teacher resources are still being translated


[PLAT-1428]: https://codedotorg.atlassian.net/browse/PLAT-1428?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
